Add Cluster creation/list

sdesimone opened the pull request:
This PR adds Cluster creation and listing, along the lines of already existing resources.

To reduce the amount of replicated code, it also introduces a new, generic `create` function and refactors existing resources' implementation to use it where appropriate.

ashenfad commented on src/bigml/api/source.clj:
```diff
@@ -35,16 +35,19 @@
    a map on failure."
   create-type)
 
+;; (defmethod create :url [url & params]
```
Leftover commented code?

ashenfad commented:
Looks good, but it'd be nice to also update the readme so it includes the new cluster resource.

sdesimone commented:
Thanks. I realised that from the perspective of the user, it is useful to have clusters *and* centroids, so I added support for that too. This entailed some more refactoring of `utils/create` since centroid creation is pretty the same as prediction's (and more of the same will come).

Besides adding details about clusters and centroids to `readme.md`, I also rewrote the `Overview` to sync it with BigML's evolution throughout the years (basically copy/pasting from the web docs).

ashenfad commented on test/bigml/test/api/source.clj:
```diff
@@ -8,11 +8,13 @@
 
 (defn- create-and-test [artifact & params]
-  (let [initial (api/get-final (apply source/create artifact params))
+  (let [{:keys [username api_key dev_mode] :as sticky-pars} params
```
Minor nit - there's no need to bind `sticky-pars` here as it will be the same as `params`.  Also, neither `username`, `api_key`, or `dev_mode` are used in the `let`.  So this could be simplified to:

```clojure
(let [sticky-pars (flatten (seq params))
      ...
```

sdesimone replied:
thanks @ashenfad. but let me just check my understanding of things a little bit...
My reasoning was this: `params` can be used to pass any source creation options; on the other hand, we only want to pass auth credentials + `dev_mode` to subsequent calls. That is why I introduced `sticky-params` the way I did. It is true that we have currently no unit tests that use creation parameters, but I think this will be the case as we grow the test suite. Does it make more sense?

jimshur replied:
@sdesimone I think your understandable confusion is with the `:as` directive, which in fact assigns the entire expression being deconstructed, not just the ones listed for the keys. Put  `(let [{:keys [a b] :as p} {:a 1 :b 2 :c 3}] p)` in the repl and see.

sdesimone replied:
Thanks, @jimshur, I see what you mean.

I have pushed a fix for that. I hope it is sufficiently idiomatic:

     (let [{:keys [username api_key dev_mode]} params
            sticky-pars [:username username :api_key api_key :dev_mode dev_mode]


> On 14 Jan 2016, at 12:11, jimshur <notifications@github.com> wrote:
> 
> In test/bigml/test/api/source.clj:
> 
> > @@ -8,11 +8,13 @@
> >    (:use clojure.test))
> >  
> >  (defn- create-and-test [artifact & params]
> > -  (let [initial (api/get-final (apply source/create artifact params))
> > +  (let [{:keys [username api_key dev_mode] :as sticky-pars} params
> 
> @sdesimone I think your understandable confusion is with the :as directive, which in fact assigns the entire expression being deconstructed, not just the ones listed for the keys. Put (let [{:keys [a b] :as p} {:a 1 :b 2 :c 3}] p) in the repl and see.
> 
> —
> Reply to this email directly or view it on GitHub.
>

jimshur replied:
Well, since, as Adam pointed out, you're not using any of the variables in the destructuring, probably just as easy to say

`(let [sticky-pars (select-keys params [:username :api_key :dev_mode])]`

sdesimone replied:
> On 14 Jan 2016, at 17:45, jimshur <notifications@github.com> wrote:
> 
> In test/bigml/test/api/source.clj:
> 
> > @@ -8,11 +8,13 @@
> >    (:use clojure.test))
> >  
> >  (defn- create-and-test [artifact & params]
> > -  (let [initial (api/get-final (apply source/create artifact params))
> > +  (let [{:keys [username api_key dev_mode] :as sticky-pars} params
> 
> Well, since, as Adam pointed out, you're not using any of the variables in the destructuring, probably just as easy to say
> 
> (let [sticky-pars (select-keys params [:username :api_key :dev_mode])]

Thanks.
I had already tried to use `select-keys`, but `params` is not a map, so that fails:
    clojure.lang.ArraySeq cannot be cast to java.util.Map

I could convert the sequence of `params` to a map, then `select-keys`, but I am not sure it is worth the effort (although `params` should seldom contain many items).

Am I missing anything?

> 
> —
> Reply to this email directly or view it on GitHub.
>

ashenfad replied:
You can use the `:as` to bind params as a hash-map in the fn definition, which makes using `select-keys` a little less messy. Like so:

```clojure
(defn bar [a & b]
  (println "Param seq: " b))

(defn foo [a & {:as b}]
  (let [c (flatten (seq (select-keys b [:x :y])))]
    (apply bar a c)))

;; user> (foo 0 :x 1 :y 2 :z 3)
;; Param seq:  (:y 2 :x 1)
```

sdesimone replied:
> On 14 Jan 2016, at 18:20, Adam Ashenfelter <notifications@github.com> wrote:
> 
> In test/bigml/test/api/source.clj:
> 
> > @@ -8,11 +8,13 @@
> >    (:use clojure.test))
> >  
> >  (defn- create-and-test [artifact & params]
> > -  (let [initial (api/get-final (apply source/create artifact params))
> > +  (let [{:keys [username api_key dev_mode] :as sticky-pars} params
> 
> You can use the :as to bind params as a hash-map in the fn definition, which makes using select-keys a little less messy. Like so:

It seems that this would be a way to use your hint:

```
(defn- create-and-test [artifact & {:as params}]
  (let [sticky-pars (flatten (seq (select-keys 
                                   params
                                   [:username :api_key :dev_mode])))
        params (flatten (seq params))
        …
```

btw, I rewrote the remote history, I hope this is not trouble since review happens on github...

ashenfad commented on readme.md:
```diff
@@ -11,28 +11,27 @@
-[PMML](http://www.dmg.org/v4-1/GeneralStructure.html)).
+`clj-bigml` aims to make it easier to use BigML.io features from
+Clojure. BigML.io features are always growing and adapting to BigML
+customers' requirements, and `clj-bigml` does currently only supports
```
and `clj-bigml` currently supports only a limited ...

ashenfad commented on readme.md:
```diff
@@ -285,12 +286,42 @@ as a demonstration.
+A [cluster](https://bigml.com/developers/clusters) is a set of groups
+of instances of a dataset that have been automatically classified
+together according to a distance measure computed using the fields of
+the dataset. Cluster are used in unsupervised learning.
```
That line is a bit awkward.  Perhaps "A clustering is an example of unsupervised learning."?

ashenfad commented on readme.md:
````diff
@@ -11,28 +11,27 @@
 For [Leiningen](https://github.com/technomancy/leiningen):
 
 ```clojure
````
If you replace this clojure formatted version block with:
`[![Clojars Project](https://img.shields.io/clojars/v/bigml/clj-bigml.svg)](https://clojars.org/bigml/clj-bigml)`

Then the version badge will automatically update in the readme whenever we have a new release.

ashenfad replied:
Oh, and we should also bump `project.clj` to version `0.2.0` in this PR.

ashenfad commented:
Thanks for updating the old readme - I forgot it was so out of date.  I've added a few more minor nits.

ashenfad commented:
Looks good, but I think we should bump the version to `0.2.0` in `project.clj`.

sdesimone commented:
> On 15 Jan 2016, at 07:52, Adam Ashenfelter <notifications@github.com> wrote:
> 
> Looks good,

Thanks.

> but I think we should bump the version to 0.2.0 in project.clj.

This is the current content of project.clj (first line):

    (defproject bigml/clj-bigml "0.2.0-SNAPSHOT"

I don’t think you mean I should remove the “SNAPSHOT”… do you?


> —
> Reply to this email directly or view it on GitHub.
>

ashenfad commented:
> I don’t think you mean I should remove the “SNAPSHOT”… do you?

Yeah, I figured we'd make this the `0.2.0` release.  But if you have more changes coming soon and want to leave it as a snapshot, that's fine too.

sdesimone commented:
ok, now it is clearer…
I will definitely make more changes (many, I hope) pretty soon, so we can leave it as it is for now.

thanks

> On 15 Jan 2016, at 17:26, Adam Ashenfelter <notifications@github.com> wrote:
> 
> I don’t think you mean I should remove the “SNAPSHOT”… do you?
> 
> Yeah, I figured we'd make this the 0.2.0 release. But if you have more changes coming soon and want to leave it as a snapshot, that's fine too.
> 
> —
> Reply to this email directly or view it on GitHub.
>

ashenfad commented:
Thanks - merged!
